File: backend/shared/go/middleware/serviceAuthMiddleware.go
package middleware

import (
	"errors"
	"net/http"
	"os"

	"github.com/Flow-Indo/LAKOO/backend/shared/go/utils"
)

const (
	ServiceAuthHeader = "x-service-auth"
	ServiceNameHeader = "x-service-name"
)

func ServiceAuthMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		token := r.Header.Get(ServiceAuthHeader)
		serviceName := r.Header.Get(ServiceNameHeader)

		if token == "" || serviceName == "" {
			utils.WriteError(w, http.StatusUnauthorized, errors.New("Service authentication required"))
			return
		}

		serviceSecret := os.Getenv("SERVICE_SECRET")
		if serviceSecret == "" {
			utils.WriteError(w, http.StatusInternalServerError, errors.New("Service secret not configured"))
			return
		}

		if err := utils.VerifyServiceToken(token, serviceSecret); err != nil {
			utils.WriteError(w, http.StatusUnauthorized, err)
			return
		}

		next.ServeHTTP(w, r)

	})
}

File: backend/shared/typescript/middleware/serviceAuthMiddleware.ts

import { Request, Response, NextFunction } from 'express';
import { verifyServiceToken } from '../utils/serviceAuth';

const SERVICE_AUTH_HEADER = 'X-Service-Auth';
const SERVICE_NAME_HEADER = 'X-Service-Name';


export const serviceAuthMiddleware = (req: Request, res: Response, next: NextFunction) => {

    const token = req.headers[SERVICE_AUTH_HEADER.toLowerCase()];
    const serviceName = req.headers[SERVICE_NAME_HEADER.toLowerCase()];

    if (!token || !serviceName) {
        return res.status(401).json({
            error: 'Service authentication required'
        });
    }

    const serviceSecret = process.env.SERVICE_SECRET;
    if (!serviceSecret) {
        return res.status(500).json({
            error: 'Service secret not configured'
        });
    }


    try {
        verifyServiceToken(token, serviceSecret)
    } catch (error) {
        return res.status(401).json({
            error: error
        })
    }
    
    //valid service auth
    next();
};
